zerver/actions/message_send.py

```diff
@@ -1523,6 +1524,57 @@ def validate_stream_id_with_pm_notification(
     return stream
 
 
+def check_can_send_private_message(
+    realm: Realm, sender: UserProfile, user_profiles: Sequence[UserProfile], recipient: Recipient
+) -> None:
+    if sender.is_bot:
+        return
+
+    for user_profile in user_profiles:
+        if not (user_profile.is_bot or user_profile.id == sender.id):
+            break
+    else:
+        return
+
+    direct_message_permission_group = realm.direct_message_permission_group
+    if (
+        not hasattr(direct_message_permission_group, "named_user_group")
+        or direct_message_permission_group.named_user_group.name != SystemGroups.EVERYONE
+    ):
+        allowed_user_ids = get_user_group_member_ids(direct_message_permission_group)
+        if sender.id not in allowed_user_ids:
+            for user_profile in user_profiles:
+                if user_profile.id in allowed_user_ids:
+                    break
+            else:
+                raise JsonableError(
+                    _("You are not allowed to send direct messages to these recipient(s).")
+                )
+
+    direct_message_initiator_group = realm.direct_message_initiator_group
+    if (
+        not hasattr(direct_message_initiator_group, "named_user_group")
+        or direct_message_initiator_group.named_user_group.name != SystemGroups.EVERYONE
+    ):
+        if is_user_in_group(direct_message_initiator_group, sender):  # nocoverage
+            return
+        if recipient.type == Recipient.PERSONAL:
+            previous_messages_exist = Message.objects.filter(
+                realm=realm,
+                recipient__type=Recipient.PERSONAL,
+                sender__in=[sender, user_profiles[0]],
+                recipient__type_id__in=[sender.id, user_profiles[0].id],
+            ).exists()
```

Comment:
This check is not correct! A message from the sender to himself would qualify.

I know there's a lot of code in a PR like this, but you need to always spend special focus on any business logic code that is making decisions to consider corner cases.

Reply:
Oh, I should have paid more attention. I have fixed it now.
```py
previous_messages_exist = (
    Message.objects.filter(
        realm=realm,
        recipient__type=Recipient.PERSONAL,
        sender__in=[sender, user_profiles[0]],
    )
    .filter(
        Q(sender=sender, recipient__type_id=user_profiles[0].id)
        | Q(sender=user_profiles[0], recipient__type_id=sender.id)
    )
    .exists()
)
```